### PR TITLE
Apply contracts using DFCC as an optional mode

### PR DESCRIPTION
**Pull request by feliperodri**
*Issue #, if available:*
N/A.

*Description of changes:*
- Apply contracts using DFCC as an optional mode;
-  Remove --unwind 1 from property check run.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by remi-delmas-3000** [comment]
Hi, initial remarks:

- DFCC should not yet be the default since it does not yet support loop contracts.
  - Could we instead introduce a `USE_DFCC` env variable  which triggers the new build steps when set ? (we would still use the existing build steps when `USE_DFCC` is not set)


When using DFCC:
- DFCC drops unused functions after instrumentation so we can remove that step;
  - moreover, dropping unused functions before contract instrumentation will drop functions that are used in contracts but in the user program, which can lead to instrumentation and analysis failures.
- DFCC havocs statics during instrumentation so we can remove that step.
- standard library configurations flags need to be passed when calling --dfcc, because that's the time when the CPROVER library gets loaded and linked to the model

**Comment by jimgrundy** [comment]
@feliperodri, should we add @karkhaz as a reviewer?

**Comment by feliperodri** [comment]
> * standard library configurations flags need to be passed when calling --dfcc, because that's the time when the CPROVER library gets loaded and linked to the model

@remi-delmas-3000 what do you mean by "standard library configurations flags"? Could you give another review?

**Review by jimgrundy — APPROVED** [review]
Looks good to me.

**Review by karkhaz — CHANGES_REQUESTED** [review]

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 1]
```diff
@@ -411,6 +411,14 @@ CBMC_CHECK_FUNCTION_CONTRACTS := $(patsubst %,--enforce-contract %, $(CHECK_FUNC
 USE_FUNCTION_CONTRACTS ?=
 CBMC_USE_FUNCTION_CONTRACTS := $(patsubst %,--replace-call-with-contract %, $(USE_FUNCTION_CONTRACTS))
 
+# Proof writers may also apply function contracts using the Dynamic Frame
+# Condition Checking (DFCC) mode. For more information on DFCC,
+# please see https://diffblue.github.io/cbmc/contracts-dev-spec-dfcc.html.
+DFCC_MODE ?= 0
+ifeq ($(DFCC_MODE),1)
```
[the remark this answers is not in the document]
Comment (karkhaz) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common:418]: (And then there's no need for `DFCC_MODE ?= 0` above)

[the remark this answers is not in the document]
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common]: Implemented as suggested. I also fixed for `LOOP_CONTRACTS`.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 2]
```diff
@@ -626,6 +634,16 @@ $(HARNESS_GOTO)2.goto: $(HARNESS_GOTO)1.goto
 
 # Fill static variable with unconstrained values
 $(HARNESS_GOTO)3.goto: $(HARNESS_GOTO)2.goto
+ifeq ($(DFCC_MODE),1)
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common]: Implemented as suggested.

File: src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common [hunk 3]
```diff
@@ -635,9 +653,20 @@ $(HARNESS_GOTO)3.goto: $(HARNESS_GOTO)2.goto
...
 $(HARNESS_GOTO)4.goto: $(HARNESS_GOTO)3.goto
+ifeq ($(DFCC_MODE),1)
+	$(LITANI) add-job \
+	  --command 'cp $^ $@' \
+	  --inputs $^ \
+	  --outputs $@ \
+	  --stdout-file $(LOGDIR)/drop_unused_functions-log.txt \
+	  --pipeline-name "$(PROOF_UID)" \
+	  --ci-stage build \
+	  --description "$(PROOF_UID): CBMC will drop unused functions during contract instrumentation"
```
[the remark this answers is not in the document]
Comment (feliperodri) [reply src/cbmc_starter_kit/template-for-repository/proofs/Makefile.common]: Implemented as suggested.

**Review by karkhaz — APPROVED** [review]
Thank you!

**Comment by remi-delmas-3000** [comment]
Other round of remarks:

when DFCC is set, the configuration flags `--malloc-may-fail`, `--malloc-fail-null` and `--string-abstraction` must be passed to `goto-instrument` at DFCC instrumentation time instead of `cbmc` at the last step of the build.

right now they are defined  like this
```
CHECKFLAGS += $(CBMC_FLAG_MALLOC_MAY_FAIL)
CHECKFLAGS += $(CBMC_FLAG_MALLOC_FAIL_NULL)
```

To support checking contracts on recursive functions, we need to add another flag

```
CHECK_FUNCTION_CONTRACTS_REC ?=
CBMC_CHECK_FUNCTION_CONTRACTS_REC := $(patsubst %,--enforce-contract-rec %, $(CHECK_FUNCTION_CONTRACTS_REC))
```

and pass that to goto-instrument.

The step
```
# Omit initialization of unused global variables (reduces problem size)
$(HARNESS_GOTO)5.goto: $(HARNESS_GOTO)4.goto
```
Should also be disabled (or even completely removed) when contracts are used (dfcc or not) because it can drop symbols that are only used in contracts and not the rest of the program, which will make instrumentation fail later.


The step
```
# Replace function contracts, check function contracts, instrument for loop contracts
$(HARNESS_GOTO)7.goto: $(HARNESS_GOTO)6.goto
	$(LITANI) add-job \
	  --command \
	    '$(GOTO_INSTRUMENT) $(CBMC_DFCC_MODE) $(CBMC_VERBOSITY) $(CBMC_CHECK_FUNCTION_CONTRACTS) $(CBMC_USE_FUNCTION_CONTRACTS) $(CBMC_APPLY_LOOP_CONTRACTS) $^ $@' \
	  --inputs $^ \
	  --outputs $@ \
	  --stdout-file $(LOGDIR)/check_function_contracts-log.txt \
	  --pipeline-name "$(PROOF_UID)" \
	  --ci-stage build \
	  --description "$(PROOF_UID): checking function contracts"
 ```

should add `--string-abstraction`, `--malloc-may-fail`, `--malloc-fail-null` if defined, as well as the new `CBMC_CHECK_FUNCTION_CONTRACTS_REC` above.

One last remark is that running coverage on the fully instrumented program might result in some perf overhead since there could be many branches in the instrumentation code that might not be covered. 

For coverage checking we might want to run a special instrumentation pass where memory allocation from preconditions and post conditions are instrumented in the program but maybe not all assigns clause and frees clause checks

**Review by remi-delmas-3000 — CHANGES_REQUESTED** [review]
We need a couple last adjustments:

when DFCC is set, the configuration flags `--malloc-may-fail`, `--malloc-fail-null` and `--string-abstraction` must be passed to `goto-instrument` at DFCC instrumentation time instead of `cbmc` at the last step of the build. Right now they are defined like this: and only get passed to `cbmc`.

```
CHECKFLAGS += $(CBMC_FLAG_MALLOC_MAY_FAIL)
CHECKFLAGS += $(CBMC_FLAG_MALLOC_FAIL_NULL)
```

To support checking contracts on recursive functions, we need to add another flag and pass that to goto-instrument:
```
CHECK_FUNCTION_CONTRACTS_REC ?=
CBMC_CHECK_FUNCTION_CONTRACTS_REC := $(patsubst %,--enforce-contract-rec %, $(CHECK_FUNCTION_CONTRACTS_REC))
```

The step:
```
# Restrict function pointers
$(HARNESS_GOTO)2.goto: $(HARNESS_GOTO)1.goto
	$(LITANI) add-job \
	  --command \
	    '$(GOTO_INSTRUMENT) $(CBMC_VERBOSITY) $(CBMC_RESTRICT_FUNCTION_POINTER) $^ $@' \
	  --inputs $^ \
	  --outputs $@ \
	  --stdout-file $(LOGDIR)/restrict_function_pointer-log.txt \
	  --pipeline-name "$(PROOF_UID)" \
	  --ci-stage build \
	  --description "$(PROOF_UID): restricting function pointers in project sources"
```

Should be omitted if `CBMC_RESTRICT_FUNCTION_POINTER` is empty (the small changes resulting from running goto-instrument without options have been shown to cause performance problems).

In general at every build step we should try to avoid calling `goto-instrument` with empty options.

The step:
```
# Omit initialization of unused global variables (reduces problem size)
$(HARNESS_GOTO)5.goto: $(HARNESS_GOTO)4.goto
	$(LITANI) add-job \
	  --command \
	    '$(GOTO_INSTRUMENT) $(CBMC_VERBOSITY) --slice-global-inits $^ $@' \
	  --inputs $^ \
	  --outputs $@ \
	  --stdout-file $(LOGDIR)/slice_global_inits-log.txt \
	  --pipeline-name "$(PROOF_UID)" \
	  --ci-stage build \
	  --description "$(PROOF_UID): slicing global initializations"
```

Should be disabled when DFCC is set (or even completely removed from the build) because it can drop symbols that are only used in contracts and not the rest of the program, which will make instrumentation fail later.

In general I think removing unused functions should be the last thing performed to cleanup the model, instead of upfront before knowing what symbols are going to be used when instrumenting contracts.

As mentioned above, the step
```
# Replace function contracts, check function contracts, instrument for loop contracts
$(HARNESS_GOTO)7.goto: $(HARNESS_GOTO)6.goto
	$(LITANI) add-job \
	  --command \
	    '$(GOTO_INSTRUMENT) $(CBMC_DFCC_MODE) $(CBMC_VERBOSITY) $(CBMC_CHECK_FUNCTION_CONTRACTS) $(CBMC_USE_FUNCTION_CONTRACTS) $(CBMC_APPLY_LOOP_CONTRACTS) $^ $@' \
	  --inputs $^ \
	  --outputs $@ \
	  --stdout-file $(LOGDIR)/check_function_contracts-log.txt \
	  --pipeline-name "$(PROOF_UID)" \
	  --ci-stage build \
	  --description "$(PROOF_UID): checking function contracts"
```

should add `--string-abstraction`, `--malloc-may-fail`, `--malloc-fail-null` if defined, as well as the new `CBMC_CHECK_FUNCTION_CONTRACTS_REC` above.

One last remark is that running coverage on the fully instrumented program might result in some perf overhead since there could be many branches in the instrumentation code that might not be covered.

For coverage checking we might want to run a special instrumentation pass where memory allocation from preconditions and post conditions are instrumented in the program but maybe not all assigns clause and frees clause checks, but that is another problem.

**Comment by feliperodri** [comment]
@remi-delmas-3000 I have implemented your latest request. Could you give another round of reviews?

> when DFCC is set, the configuration flags `--malloc-may-fail`, `--malloc-fail-null` and `--string-abstraction` must be passed to `goto-instrument` at DFCC instrumentation time instead of `cbmc` at the last step of the build. Right now they are defined like this: and only get passed to `cbmc`.
> 
> ```
> CHECKFLAGS += $(CBMC_FLAG_MALLOC_MAY_FAIL)
> CHECKFLAGS += $(CBMC_FLAG_MALLOC_FAIL_NULL)
> ```

We don't currently support `--string-abstraction` in our starter-kit, so I'm leaving this one out of this change. We now pass the configurations flags along with the DFCC mode.
 
> To support checking contracts on recursive functions, we need to add another flag and pass that to goto-instrument:
> 
> ```
> CHECK_FUNCTION_CONTRACTS_REC ?=
> CBMC_CHECK_FUNCTION_CONTRACTS_REC := $(patsubst %,--enforce-contract-rec %, $(CHECK_FUNCTION_CONTRACTS_REC))
> ```

☝️ Implemented as suggested.

> The step:
> 
> ```
> # Restrict function pointers
> $(HARNESS_GOTO)2.goto: $(HARNESS_GOTO)1.goto
> 	$(LITANI) add-job \
> 	  --command \
> 	    '$(GOTO_INSTRUMENT) $(CBMC_VERBOSITY) $(CBMC_RESTRICT_FUNCTION_POINTER) $^ $@' \
> 	  --inputs $^ \
> 	  --outputs $@ \
> 	  --stdout-file $(LOGDIR)/restrict_function_pointer-log.txt \
> 	  --pipeline-name "$(PROOF_UID)" \
> 	  --ci-stage build \
> 	  --description "$(PROOF_UID): restricting function pointers in project sources"
> ```
> 
> Should be omitted if `CBMC_RESTRICT_FUNCTION_POINTER` is empty (the small changes resulting from running goto-instrument without options have been shown to cause performance problems).

☝️ Implemented as suggested.

> In general at every build step we should try to avoid calling `goto-instrument` with empty options.

I'm gonna restrict this PR to only change code necessary to make our Makefile.common support DFCC soundly. I'll create a GitHub issue to track this.

> The step:
> 
> ```
> # Omit initialization of unused global variables (reduces problem size)
> $(HARNESS_GOTO)5.goto: $(HARNESS_GOTO)4.goto
> 	$(LITANI) add-job \
> 	  --command \
> 	    '$(GOTO_INSTRUMENT) $(CBMC_VERBOSITY) --slice-global-inits $^ $@' \
> 	  --inputs $^ \
> 	  --outputs $@ \
> 	  --stdout-file $(LOGDIR)/slice_global_inits-log.txt \
> 	  --pipeline-name "$(PROOF_UID)" \
> 	  --ci-stage build \
> 	  --description "$(PROOF_UID): slicing global initializations"
> ```
> 
> Should be disabled when DFCC is set (or even completely removed from the build) because it can drop symbols that are only used in contracts and not the rest of the program, which will make instrumentation fail later.

☝️ Implemented as suggested.

> In general I think removing unused functions should be the last thing performed to cleanup the model, instead of upfront before knowing what symbols are going to be used when instrumenting contracts.

I'm gonna restrict this PR to only change code necessary to make our Makefile.common support DFCC soundly. I'll create a GitHub issue to track this.

> One last remark is that running coverage on the fully instrumented program might result in some perf overhead since there could be many branches in the instrumentation code that might not be covered.
> 
> For coverage checking we might want to run a special instrumentation pass where memory allocation from preconditions and post conditions are instrumented in the program but maybe not all assigns clause and frees clause checks, but that is another problem.

I'm gonna restrict this PR to only change code necessary to make our Makefile.common support DFCC soundly. I'll create a GitHub issue to track this.

**Review by remi-delmas-3000 — COMMENTED** [review]
- the `malloc` failure mode flags will be both passed along with `--dfcc` and to `cbmc` when DFCC is used (which should have no effect but feels kind of weird)
- I forgot to tell you initially but the `$(NONDET_STATIC)` commands, in particular the `--nondet-static-exclude VAR` flags need to be passed to DFCC as well, since DFCC nondet-initialises the model too.
- last, we need a way to pass extra options like `--string-abstraction` to DFCC. Before it was passed along the rest in `CHECKFLAGS` to cbmc at the end but this really is a library configuration flag

**Comment by feliperodri** [comment]
> * the `malloc` failure mode flags will be both passed along with `--dfcc` and to `cbmc` when DFCC is used (which should have no effect but feels kind of weird)

I though it was simple to call CBMC with all `malloc` flags again, rather than striping it out from the command if DFCC mode is on, since it has no effect and, in the future, we want to make DFCC as default.

>     * I forgot to tell you initially but the `$(NONDET_STATIC)` commands, in particular the `--nondet-static-exclude VAR` flags need to be passed to DFCC as well, since DFCC nondet-initialises the model too.

👆 Implemented as suggested.

>     * last, we need a way to pass extra options like `--string-abstraction` to DFCC. Before it was passed along the rest in `CHECKFLAGS` to cbmc at the end but this really is a library configuration flag

👆 Implemented as suggested.

**Review by remi-delmas-3000 — APPROVED** [review]

**Pull request merged**